pkg/apis/scheduling/types.go

```diff
@@ -88,3 +88,128 @@ type PriorityClassList struct {
 	// items is the list of PriorityClasses.
 	Items []PriorityClass
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Workload allows for expressing scheduling constraints that should be used
+// when managing lifecycle of workloads from scheduling perspective,
+// including scheduling, preemption, eviction and other phases.
+type Workload struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	//
+	// +optional
+	metav1.ObjectMeta
+
+	// Spec defines the desired behavior of a Workload.
+	//
+	// +required
+	Spec WorkloadSpec
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// WorkloadList contains a list of Workload resources.
+type WorkloadList struct {
+	metav1.TypeMeta
+	// Standard list metadata.
+	//
+	// +optional
+	metav1.ListMeta
+
+	// Items is the list of Workloads.
+	Items []Workload
+}
+
+// WorkloadMaxPodGroups is the maximum number of pod groups per Workload.
+const WorkloadMaxPodGroups = 8
+
+// WorkloadSpec defines the desired state of a Workload.
+type WorkloadSpec struct {
+	// ControllerRef is an optional reference to the controlling object, such as a
+	// Deployment or Job. This field is intended for use by tools like CLIs
+	// to provide a link back to the original workload definition.
+	// When set, it cannot be changed.
+	//
+	// +optional
+	ControllerRef *TypedLocalObjectReference
+
+	// PodGroups is the list of pod groups that make up the Workload.
+	// The maximum number of pod groups is 8. This field is immutable.
+	//
+	// +required
+	// +listType=map
+	// +listMapKey=name
+	PodGroups []PodGroup
+}
+
+// TypedLocalObjectReference allows to reference typed object inside the same namespace.
+type TypedLocalObjectReference struct {
+	// APIGroup is the group for the resource being referenced.
+	// If APIGroup is empty, the specified Kind must be in the core API group.
+	// For any other third-party types, setting APIGroup is required.
+	// It must be a DNS subdomain.
+	//
+	// +optional
+	APIGroup string
+	// Kind is the type of resource being referenced.
+	// It must be a path segment name.
+	//
+	// +required
+	Kind string
+	// Name is the name of resource being referenced.
+	// It must be a path segment name.
+	//
+	// +required
+	Name string
+}
+
+// PodGroup represents a set of pods with a common scheduling policy.
+type PodGroup struct {
+	// Name is a unique identifier for the PodGroup within the Workload.
+	// It must be a DNS label. This field is immutable.
+	//
+	// +required
+	Name string
+
+	// Policy defines the scheduling policy for this PodGroup.
+	//
+	// +required
+	Policy PodGroupPolicy
+}
+
+// PodGroupPolicy defines the scheduling configuration for a PodGroup.
+type PodGroupPolicy struct {
+	// Basic specifies that the pods in this group should be scheduled using
```

Comment:
Don't we usually use a discriminator like a `type` field for enums, especially if there are members that have no settings? E.G. in `deployment.spec.strategy` there is a `type`

Reply:
Its actually like that in the kep: https://github.com/kubernetes/enhancements/blob/363d884be5f231049767237926dfade5febdd4ca/keps/sig-scheduling/4671-gang-scheduling/README.md#L355